joblib/numpy_pickle.py

```diff
@@ -1,229 +1,251 @@
+        We added some code to support versions of numpy prior to 1.9.
+        """
+        # Set buffer size to 16 MiB to hide the Python loop overhead.
+        buffersize = max(16 * 1024 ** 2 // array.itemsize, 1)
+        np_ver = [int(x) for x in pickler.np.__version__.split('.', 2)[:2]]
+
+        if array.dtype.hasobject:
+            # We contain Python objects so we cannot write out the data
+            # directly. Instead, we will pickle it out with version 2 of the
+            # pickle protocol.
+            pickle.dump(array, pickler.file, protocol=2)
+        elif self.order == 'F':
+            if pickler.np.compat.isfileobj(pickler.file):
+                array.T.tofile(pickler.file)
+            else:
+                # 'tobytes' is available since numpy 1.9.
+                if np_ver[0] >= 1 and np_ver[1] >= 9:
+                    for chunk in pickler.np.nditer(array,
+                                                   flags=['external_loop',
+                                                          'buffered',
+                                                          'zerosize_ok'],
+                                                   buffersize=buffersize,
+                                                   order='F'):
+                        pickler.file.write(chunk.tobytes('C'))
+                else:
+                    pickler.file.write(array.T.tostring('C'))
+        else:
+            if pickler.np.compat.isfileobj(pickler.file):
+                array.tofile(pickler.file)
+            else:
+                # 'tobytes' is available since numpy 1.9.
+                if np_ver[0] >= 1 and np_ver[1] >= 9:
+                    for chunk in pickler.np.nditer(array,
+                                                   flags=['external_loop',
+                                                          'buffered',
+                                                          'zerosize_ok'],
+                                                   buffersize=buffersize,
+                                                   order='C'):
+                        pickler.file.write(chunk.tobytes('C'))
+                else:
+                    pickler.file.write(array.tostring('C'))
+
+    def read_array(self, unpickler):
+        """Read array from unpickler file handle.
+
+        This function is an adapation of the numpy read_array function
+        available in version 0.10 in numpy/lib/format.py.
+        """
+        if len(self.shape) == 0:
+            count = 1
+        else:
+            count = unpickler.np.multiply.reduce(self.shape)
+        # Now read the actual data.
+        if self.dtype.hasobject:
+            # The array contained Python objects. We need to unpickle the data.
+            array = pickle.load(unpickler.file_handle)
+        else:
+            if (unpickler.np.compat.isfileobj(unpickler.file_handle) and
+                    count*self.dtype.itemsize/1e6 > 700):
```

Comment:
Here is what I did:

```python
import os
import gc

import numpy as np

import joblib

%load_ext memory_profiler

rng = np.random.RandomState(0)

                      
sizes = np.arange(100, 1000, 100)
filenames = ['file_{}_MiB.raw'.format(size) for size in sizes]

def maybe_create_files():
    files_missing = [fn for fn in filenames if not os.path.exists(fn)]
    if files_missing:
        array = rng.randn(max(sizes) * 1024**2 // 8).astype('float64')
        for filename, size in zip(filenames, sizes):
            max_index = size * 1024**2 // 8
            array[:max_index].tofile(filename)

        del array


maybe_create_files()

for filename, size in zip(filenames, sizes):
    with open(filename, 'rb') as fileobj:
        print('size: {} MiB'.format(size))
        %memit -i 0.01 np.fromfile(fileobj)
```

Output:
```
size: 100 MiB
peak memory: 115.89 MiB, increment: 82.85 MiB
size: 200 MiB
peak memory: 203.93 MiB, increment: 172.58 MiB
size: 300 MiB
peak memory: 327.93 MiB, increment: 296.57 MiB
size: 400 MiB
peak memory: 419.93 MiB, increment: 388.57 MiB
size: 500 MiB
peak memory: 519.93 MiB, increment: 488.57 MiB
size: 600 MiB
peak memory: 619.93 MiB, increment: 588.57 MiB
size: 700 MiB
peak memory: 715.93 MiB, increment: 684.57 MiB
size: 800 MiB
peak memory: 801.93 MiB, increment: 770.57 MiB
size: 900 MiB
peak memory: 909.93 MiB, increment: 878.57 MiB
```

From what I can gather there seems to be some kind of overhead around 20-30MB, I am guessing from a buffer somewhere.